lib/tesla/middleware/fuse.ex

```diff
@@ -48,30 +55,36 @@ if Code.ensure_loaded?(:fuse) do
     @impl Tesla.Middleware
     def call(env, next, opts) do
       opts = opts || []
-      name = Keyword.get(opts, :name, env.__module__)
 
-      case :fuse.ask(name, :sync) do
+      context = %{
+        name: Keyword.get(opts, :name, env.__module__),
+        should_melt: Keyword.get(opts, :should_melt, &match?({:error, _}, &1))
+      }
+
+      case :fuse.ask(context.name, :sync) do
         :ok ->
-          run(env, next, name)
+          run(env, next, context)
 
         :blown ->
           {:error, :unavailable}
 
         {:error, :not_found} ->
-          :fuse.install(name, Keyword.get(opts, :opts, @defaults))
-          run(env, next, name)
+          :fuse.install(context.name, Keyword.get(opts, :opts, @defaults))
+          run(env, next, context)
       end
     end
 
-    defp run(env, next, name) do
-      case Tesla.run(env, next) do
-        {:ok, env} ->
-          {:ok, env}
+    defp run(env, next, context) do
+      Tesla.run(env, next)
+      |> maybe_melt_fuse(context)
+    end
 
-        {:error, _reason} ->
-          :fuse.melt(name)
-          {:error, :unavailable}
+    defp maybe_melt_fuse(res, context) do
+      if context.should_melt.(res) do
+        :fuse.melt(context.name)
       end
+
+      res
```

Comment:
Why do we need a separate function here? How about 

```elixir
def run(env, next, %{should_melt: should_melt, name: name}) do
  res = Tesla.run(env, next)

  if should_melt.(res), do: :fuse.melt(name)

  res
end
```

?

Reply:
Sure, not really needed. Indeed, your suggestion was my initial implementation xD

Let me do it this way :)